passkey: verify assertion

ikerexxe opened the pull request:
Include the assertion verification in the passkey helper process. For that purpose, iterate all the passkey devices connected to the system, prepare the assertion request data, decode the public key, get the authenticator data, get the device options and compare them with the organization policy and finally verify the assert.

Resolves: https://github.com/SSSD/sssd/issues/6228

## How to test

The words in bold should be replaced by a real value.

Generate the key registration:
$ ./passkey_child --register --username=**user** --domain=**test.com**
Enter PIN: 
Please touch the device.
passkey:jYXUcpMDflraOutYQ34tesRq7m3QSthrsuPLfNm+vPIJZVwTQf6BcQg935SouNqRevu4vVJI73inw7qI3ICFAw==,yGGMi6liZ8Y2m841WT0cI9Mfd3jE5H08hWKlRqvK3aDUn3vi27kbBKtbCnAoJri3vB/fZ4xBW+Kz2/yUYNbbyQ==

The following options can also be set when executing the previous command to print debugging logs
--debug-level=9 --logger=stderr

Authenticate the user. The output from the previous command should be used as an input for this one (key handle and public key). Remember that you'll need to touch the device, input the PIN and touch the device again, as the helper process is supposed to be called from SSSD and the prompts will be handled there.
$ ./passkey_child --authenticate --username=**user** --domain=**test.com** --key-handle=**jYXUcpMDflraOutYQ34tesRq7m3QSthrsuPLfNm+vPIJZVwTQf6BcQg935SouNqRevu4vVJI73inw7qI3ICFAw==** --public-key=**yGGMi6liZ8Y2m841WT0cI9Mfd3jE5H08hWKlRqvK3aDUn3vi27kbBKtbCnAoJri3vB/fZ4xBW+Kz2/yUYNbbyQ==**

ikerexxe commented on src/tests/cmocka/test_passkey_child.c:
```diff
@@ -92,6 +100,12 @@ static int teardown(void **state)
  * WRAPPERS
  **********************/
+unsigned int
+__wrap_sleep(unsigned int seconds)
```
False positive, it's being used.

ikerexxe commented:
The COPR builds are [compiling correctly](https://copr.fedorainfracloud.org/coprs/g/sssd/pr6384/builds/) but there seems to be an error reporting the status.

justin-stephenson commented:
> 
> Authenticate the user. The output from the previous command should be used as an input for this one (key handle and public key).

In the local (non-centralized) authentication case, where will the key handle and public key be read from? Sorry if I missed it somewhere.

alexey-tikhonov commented:
> In the local (non-centralized) authentication case, where will the key handle and public key be read from? Sorry if I missed it somewhere.

Most probably SSSD isn't going to support this use case.

justin-stephenson commented:
When running the authenticate operation with `--debug-level=9 --logger=stderr`, I see some useful log messages until my device blinks to touch it. From then on I only see the following logs when touching the device, entering the PIN and touching the device again.

~~~
[passkey_child[430188]] [get_device_assert] (0x0200): Assertion found in FIDO2 key 0.
Enter PIN: 
~~~

IMO it would be nice to log a debug message to indicate what is happening, and possibly what is the passkey child waiting for. And also a 'success' authentication message when done, then it will be easy to see when the child returns to the PAM responder.

justin-stephenson commented on src/passkey_child/passkey_child_common.c:
```diff
@@ -319,3 +369,99 @@ register_key(struct passkey_data *data)
 }
+
+errno_t
+authenticate(struct passkey_data *data)
```
high level description of what this function does could be mentioned here in a comment here, like what is in the commit message.

ikerexxe replied:
I'll add that to the `passkey_child.h` since that is where I am adding all the comments.

ikerexxe replied:
Added to `passkey_child.h`

justin-stephenson commented on src/passkey_child/passkey_child_common.c:
```diff
@@ -319,3 +369,99 @@ register_key(struct passkey_data *data)
+
+        ret = verify_assert(data, dev, assert);
+        if (ret == FIDO_OK) {
+            // End the loop only if the authentication has been successful
```
sssd coding style states Do not use C99 line comments (// line comment)

ikerexxe replied:
Fixed

justin-stephenson commented on src/passkey_child/passkey_child_assert.c:
```diff
@@ -0,0 +1,270 @@
+
+    has_pin = fido_dev_has_pin(dev);
+    if (has_pin == true) {
+        pin_len = read_pin(&pin);
```
For my knowledge, how will this be changed when SSSD handles the prompting? just this block of code will be removed?

ikerexxe replied:
I've removed it and updated the instructions to take into account that the PIN is an input to the child process

ikerexxe commented:
> > Authenticate the user. The output from the previous command should be used as an input for this one (key handle and public key).
> 
> In the local (non-centralized) authentication case, where will the key handle and public key be read from? Sorry if I missed it somewhere.

This use case won't be supported.

> IMO it would be nice to log a debug message to indicate what is happening, and possibly what is the passkey child waiting for.

All the prompting will be handled by pam_sss. Thanks for noticing it.

This means that the PIN should be an input to the child, which in turn means that pam_sss should ask for it. I've added the option and updated the testing instructions.

It also means that any message warning the user that it should touch the device when it blinks should also be printed in pam_sss (`touch-prompt` sssd.conf option). Taking into account the workflow (touching the device twice isn't obvious) I think that we should change the default value of `touch` to `true`and `touch-prompt` to `Please touch the device when it blinks. The device may blink twice` or something similar.

> And also a 'success' authentication message when done, then it will be easy to see when the child returns to the PAM responder.

Done.

justin-stephenson commented on src/passkey_child/passkey_child_common.c:
```diff
@@ -319,3 +323,99 @@ register_key(struct passkey_data *data)
+
+    for (int i = 0; i < dev_number; i++) {
+        DEBUG(SSSDBG_TRACE_FUNC,
+              "Attempting authentication with device number %d.\n", i + 1);
```
Not necessary but I thought it could be useful to have some helpful log messages here, something like below.

~~~
         DEBUG(SSSDBG_TRACE_FUNC,
               "Attempting authentication with device number %d.\n", i + 1);
 
+        DEBUG(SSSDBG_TRACE_FUNC, "Decoding public key\n");
         ret = decode_public_key(data);
         if (ret != FIDO_OK) {
             goto done;
         }
 
+        DEBUG(SSSDBG_TRACE_FUNC, "Obtaining device assertation\n");
         ret = get_device_assert(data, dev_list, dev_number, assert,
                                 &dev);
         if (ret != FIDO_OK) {
             goto done;
         }
 
+        DEBUG(SSSDBG_TRACE_FUNC, "Getting device options\n");
         ret = get_device_options(data, dev);
         if (ret != FIDO_OK) {
             goto done;
         }
 
+        DEBUG(SSSDBG_TRACE_FUNC, "Setting assert options\n");
         ret = set_assert_options(data, assert);
         if (ret != FIDO_OK) {
             DEBUG(SSSDBG_OP_FAILURE, "Failed to reset assert options.\n");
             goto done;
         }
 
+        DEBUG(SSSDBG_TRACE_FUNC, "Setting assert client data hash\n");
         ret = set_assert_client_data_hash(assert);
         if (ret != FIDO_OK) {
             DEBUG(SSSDBG_OP_FAILURE, "Failed to reset client data hash.\n");
             goto done;
         }
 
+        DEBUG(SSSDBG_TRACE_FUNC, "Verify assert\n");
         ret = verify_assert(data, dev, assert);
         if (ret == FIDO_OK) {
             /* End the loop only if the authentication has been successful */
~~~

ikerexxe replied:
Good idea! I've added them to the updated patch.

ikerexxe commented:
> By the way, is it really necessary to require touching the device twice?

Yes, it's necessary. The first round of blinking is used to select the key to be used to obtain the device assertion for the authentication. If you have more than one key connected to the computer then you see that one of the keys starts blinking for some seconds, and if you don't touch it, then the second key starts blinking. The second round of blinking is used to verify the assertion.

If you are interested these would be the flow:
```
...
[passkey_child[40650]] [authenticate] (0x0400): Decoding public key.
[passkey_child[40650]] [authenticate] (0x0400): Obtaining device assertation.
[passkey_child[40650]] [get_device_assert] (0x0400): Working with 2 authenticator(s).
```
First key starts blinking for several seconds
```
[passkey_child[40650]] [get_device_assert] (0x0200): Assertion not found in FIDO2 key 0.
```
First key stops blinking and the second one starts
```
[passkey_child[40650]] [get_device_assert] (0x0200): Assertion not found in FIDO2 key 1.
[passkey_child[40650]] [get_device_assert] (0x0040): Assertion not found.
```

ikerexxe commented:
@LDVG can you also review this PR?

LDVG commented on src/passkey_child/passkey_child_assert.c:
```diff
@@ -0,0 +1,261 @@
+{
+    errno_t ret;
+
+    ret = fido_assert_set_uv(assert, data->user_verification);
```
In the pre-flight assertion, the device options have not yet been matched to the device. If `data->user_verification` is `FIDO_OPT_FALSE` and the authenticator does not support built-in user verification, this may cause the pre-flight to fail with `CTAP2_ERR_INVALID_OPTION` even if the authenticator knows of the credential.

ikerexxe replied:
I think this is already fixed with the pre-flight assertion but take a look just in case.

LDVG commented:
> > By the way, is it really necessary to require touching the device twice?
> 
> Yes, it's necessary.

This is not necessarily true. Credentials with a protection level of `userVerificationOptional` (the default for most authenticators) or `userVertificationOptionalWithCredentialIDList` do not require UP to generate an assertion. This is leveraged for so called "pre-flight" assertions with UP=false; UV=omit/false; PIN=null, and dummy data to determine which of the user's connected authenticators hold a particular credential before following up the real assertion.

For authenticators with user verification configured and with a credential protection level `userVerificationRequired` (or if you're using discoverable credentials and do not pass credential ID) it's another story. Other clients, such as browsers, usually solve this by making you select an authenticator via touch (e.g. `fido_dev_get_touch_{begin,status}()`) before issuing  the assertion request depending on the Relying Party's user verification preferences.

See further notes in-line to the code.

ikerexxe commented on src/passkey_child/passkey_child.c:
```diff
@@ -70,7 +70,14 @@ int main(int argc, const char *argv[])
         }
     } else if (data.action == ACTION_AUTHENTICATE) {
-        ERROR("This action isn't implemented yet.\n");
+        ret = authenticate(&data);
```
Yes, that's right.

LDVG commented on src/passkey_child/passkey_child_devices.c:
```diff
@@ -84,3 +84,143 @@ select_device(fido_dev_info_t *dev_list, size_t dev_index, fido_dev_t *dev)
+        }
+    }
+
+    if (data->user_verification == FIDO_OPT_FALSE && has_uv < 0) {
+        data->user_verification = FIDO_OPT_OMIT;
+    }
+
+    if (data->user_verification == FIDO_OPT_TRUE && has_pin != true
+        && has_uv != true) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "Policy enabled user-verification but there isn't any "
+              "verification method set.\n");
+        ret = EINVAL;
+        goto done;
+    }
```
Yes, those functions return booleans. However, your variables in the current code above are `int`. :-)

ikerexxe replied:
Fixed.

ikerexxe commented on src/passkey_child/passkey_child_devices.c:
```diff
@@ -84,3 +84,143 @@ select_device(fido_dev_info_t *dev_list, size_t dev_index, fido_dev_t *dev)
+        goto done;
+    }
+
+    if (data->user_verification == FIDO_OPT_FALSE
+        && (has_uv == true || has_pin == true)) {
+        DEBUG(SSSDBG_CONF_SETTINGS,
+              "Policy disabled user-verification but the key settings are "
+              "enforcing it. Thus, enabling user-verification.\n");
+        data->user_verification = FIDO_OPT_TRUE;
+    }
```
mmm it's set to FIDO_OPT_TRUE, so I don't know if you were confused. Can you review this comment?

LDVG replied:
Calling this process with `--user-verification=false`, with an authenticator that has either a PIN or built-in UV set, ultimately results in `data->user_verification` being set to `FIDO_OPT_TRUE`. Correct?

Per the debug message, you're suggesting that the authenticator required UV. However, this is not necessarily true. You can request the assertion without UV (unless the `credProtect` extension was used with `userVerificationRequired` when the credential was created).

What I meant by `FIDO_OPT_OMIT` is that if the calling process has omitted `--user-verification` you might prefer enabling user verification if the authenticator supports it, e.g. `if (data->user_verification == FIDO_OPT_OMIT) ...`.

ikerexxe replied:
Fixed.

ikerexxe commented on src/passkey_child/passkey_child_common.c:
```diff
@@ -86,7 +88,9 @@ parse_arguments(int argc, const char *argv[], struct passkey_data *data)
          _("Domain"), NULL},
-        {"public-key", 0, POPT_ARG_STRING, &data->public_key, 0,
+        {"pin", 0, POPT_ARG_STRING, &data->pin, 0,
+         _("Passkey PIN"), NULL},
```
The calling process will prompt for the PIN when the user is authenticating and it's up to the user to provide it.

LDVG replied:
How will the calling process know that the authenticator will require a PIN?

ikerexxe replied:
Unfortunately it doesn't know it. It's the user who does. So, I think we'll need to ask for the PIN always.

ikerexxe replied:
I'm keeping this point open as I need to think this through.

LDVG replied:
I'd additionally like to make another point on this topic:

Assuming that ...
- the user has registered multiple authenticators;
- they have set different PINS on each of these authenticators; and
- have some/all plugged in during authentication.

Since you're using pre-flights, the user cannot know which authenticator will be selected by the algorithm. It is entirely dependent on which authenticator that happens to be enumerated first. They are then forced to guess which authenticator was selected first and risk burning a PIN retry.

This is why it might be preferable to have the user select an authenticator via touch whenever UV is requested by the RP.

LDVG replied:
Another data point: passing the PIN as an argument is leaky (visible to other users monitoring processes on the system).

ikerexxe replied:
> Since you're using pre-flights, the user cannot know which authenticator will be selected by the algorithm. It is entirely dependent on which authenticator that happens to be enumerated first. They are then forced to guess which authenticator was selected first and risk burning a PIN retry.

This is a limitation that we are willing to risk. pam_sss will be in charge of getting the PIN and inputting it to the helper process.

> Another data point: passing the PIN as an argument is leaky/a potential security issue (visible to other users monitoring processes on the system).

This is a more serious concern and I will bring it to our meeting.

alexey-tikhonov replied:
> > Another data point: passing the PIN as an argument is leaky/a potential security issue (visible to other users monitoring processes on the system).
> 
> This is a more serious concern and I will bring it to our meeting.

`p11_child` receives PIN via fd:
https://github.com/SSSD/sssd/blob/955192b12fafdb55daaa0b12381322031fd20816/src/p11_child/p11_child_common.c#L377

ikerexxe replied:
> Another data point: passing the PIN as an argument is leaky/a potential security issue (visible to other users monitoring processes on the system).

I followed Alexey's suggestion and this is read from stdin to avoid the security issue that you reported.

LDVG replied:
> I followed Alexey's suggestion and this is read from stdin to avoid the security issue that you reported.

:+1:

IIUC, this still means that the calling process needs to supply a PIN every time UV is requested, otherwise the process may hang indefinitely?

> This is a limitation that we are willing to risk. pam_sss will be in charge of getting the PIN and inputting it to the helper process. For that purpose pam_sss will use a PAM conversation to communicate with the user. There isn't any other way of doing it if we want to enable the authentication by different means (console, graphical interface, etc.)

I'm assuming the limitation you're referring to here is the speculative PIN prompt? There are multiple ways you could work around this; employing a simple IPC or have the passkey child return a particular error status when it needs a PIN to name a few. Both would make `pam_sss`'s PIN prompts reactive instead and users would not have to always type their PIN. Consider for example a valid fingerprint match with the Yubikey Bio. Like previously mentioned, I won't block this PR, but I highly encourage you to consider these improvements to the UX. :slightly_smiling_face:

ikerexxe replied:
> IIUC, this still means that the calling process needs to supply a PIN every time UV is requested, otherwise the process may hang indefinitely?

Yes, but take into account that sssd's main process is the caller, and it'll be aware of this behaviour,

> I'm assuming the limitation you're referring to here is the speculative PIN prompt? There are multiple ways you could work around this: employing a simple IPC, signals, or having the passkey child return a particular error status when it needs a PIN, to name a few. These would make `pam_sss`'s PIN prompts reactive instead and users would not have to always type their PIN. Consider for example a valid fingerprint match with the Yubikey Bio. Like previously mentioned, I won't block this PR, but I highly encourage you to consider these improvements to the UX at some point. slightly_smiling_face

For the moment I'll keep the PR as it is, and I'll see how to improve the UX.

LDVG commented on src/passkey_child/passkey_child_common.c:
```diff
@@ -319,3 +323,105 @@ register_key(struct passkey_data *data)
+        }
+
+        DEBUG(SSSDBG_TRACE_FUNC, "Comparing the device and policy options.\n");
+        ret = get_device_options(data, dev);
```
This function mutates `data` and no copy of the original is kept. This will have unintended side-effects if there are more than one authenticator connected to the system and the outer loop runs more than once.

ikerexxe replied:
I think this doesn't apply any longer.

ikerexxe commented on src/passkey_child/passkey_child.h:
```diff
@@ -31,6 +31,7 @@
 #define DEFAULT_CUE "Please touch the device."
 
 #define DEVLIST_SIZE    64
+#define AUTHLIST_SIZE   64
```
Removed

ikerexxe commented on src/passkey_child/passkey_child_common.c:
```diff
@@ -319,3 +323,105 @@ register_key(struct passkey_data *data)
+        goto done;
+    }
+
+    for (int i = 0; i < dev_number; i++) {
```
I've implemented the pre-flight assertion method, please take a look at it and provide feedback.

ikerexxe commented on src/passkey_child/passkey_child_common.c:
```diff
@@ -319,3 +323,105 @@ register_key(struct passkey_data *data)
+        ret = decode_public_key(data);
+        if (ret != FIDO_OK) {
+            goto done;
+        }
```
I forgot to add the possibility of inputting multiple public keys, so I've added a new commit to cover that point.

Apart from that I also removed the public key decoding for any loop and now it's only done once.

ikerexxe commented on src/passkey_child/passkey_child_devices.c:
```diff
@@ -84,3 +84,143 @@ select_device(fido_dev_info_t *dev_list, size_t dev_index, fido_dev_t *dev)
+                has_uv = value[i];
+            }
+        }
+    }
```
Updated.

ikerexxe commented on src/passkey_child/passkey_child_common.c:
```diff
@@ -319,3 +323,105 @@ register_key(struct passkey_data *data)
+            goto done;
+        }
+
+        DEBUG(SSSDBG_TRACE_FUNC, "Setting assert client data.\n");
```
Fixed.

ikerexxe commented on src/passkey_child/passkey_child_assert.c:
```diff
@@ -0,0 +1,261 @@
+}
+
+errno_t
+verify_assert(struct passkey_data *data, fido_dev_t *dev,
```
I've separated the function in two as you suggested.

ikerexxe commented on src/passkey_child/passkey_child_assert.c:
```diff
@@ -0,0 +1,261 @@
+                  ret, fido_strerr(ret));
+            goto done;
+        }
+    }
```
The user-verification setting has been moved to another function and it is set according to the policy/device.

ikerexxe commented:
> This is not necessarily true. Credentials with a protection level of `userVerificationOptional` (the default for most authenticators) or `userVertificationOptionalWithCredentialIDList` typically do not require _any_ user interaction to generate an assertion as long as you provide the Credential ID(s), which is required for non-discoverable credentials. This is leveraged for so called "pre-flight" assertions with UP=false; UV=omit/false; PIN=null, used to determine which of the user's connected authenticators hold a particular credential before following up with the real assertion.
> 
> For credentials with a credential protection level of `userVerificationRequired` it's another story. See further notes in-line to the code.
> 
> References:
> 
>     * https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-20210615.html#pre-flight
> 
>     * https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-20210615.html#sctn-credProtect-extension

I implemented the pre-flight method and the key only needs to be touched once. Thanks for the feedback.

ikerexxe commented on src/tests/cmocka/test_passkey_child.c:
```diff
@@ -257,64 +271,190 @@
+}
+
+size_t
+__wrap_fido_cbor_info_options_len(const fido_cbor_info_t *ci)
```
Removed.

ikerexxe commented on src/tests/cmocka/test_passkey_child.c:
```diff
@@ -257,64 +271,190 @@
+}
+
+char **
+__wrap_fido_cbor_info_options_name_ptr(const fido_cbor_info_t *ci)
```
Removed.

ikerexxe commented on src/tests/cmocka/test_passkey_child.c:
```diff
@@ -257,64 +271,190 @@
+}
+
+const bool *
+__wrap_fido_cbor_info_options_value_ptr(const fido_cbor_info_t *ci)
```
Removed.

LDVG commented on src/passkey_child/passkey_child_devices.c:
```diff
@@ -55,32 +55,166 @@ list_devices(fido_dev_info_t *dev_list, size_t *dev_number)
-    if (ret != FIDO_OK) {
-        DEBUG(SSSDBG_OP_FAILURE, "fido_dev_open failed [%d]: %s.\n",
-              ret, fido_strerr(ret));
+    if (_data->user_verification == FIDO_OPT_FALSE
+        && (has_uv == true || has_pin == true)) {
+        DEBUG(SSSDBG_CONF_SETTINGS,
+              "Policy disabled user-verification but the key settings are "
+              "enforcing it. Thus, delegating the decission to the key.\n");
+        _data->user_verification = FIDO_OPT_OMIT;
     }
```
The key is likely not enforcing anything here (unless `alwaysUv` is configured). Omitting `uv` only means that no built-in UV will be performed in most cases. To honor the Relying Party's request, you probably want to keep this set to `false`. What need to decide though, is what your client prefers to do when the Relying Party has not specified any policy for user verification (`_data->user_verification == FIDO_OPT_OMIT`), if that is something you're going to support.

LDVG replied:
> Is that possible? Let's see it with an example, if the key has a PIN set, would it be possible to request the assert without providing a PIN?

This is the default behavior, yes.

There are some exceptions:
- the credential is protected with[ `userVerificationRequired`](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-20210615.html#op-getassn-step-locate-credentials) and the pre-flight algorithm will never discover this credential (as previously discussed); and
- the authenticator has [`alwaysUv`](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-20210615.html#ref-for-getinfo-alwaysuv%E2%91%A0) configured and either (silently) promotes the request to use built-in UV (if possible) or returns an error.

I'm not aware of any authenticator that creates credentials with `userVerificationRequired` unsolicited. The YubiKey Bio has `alwaysUv` enabled by default.

> Yes, that would be possible. My idea was to rely on the keys settings, but taking into account your previous statement I don't know if that's possible.

I think that makes sense for the omit case. If the RP has no preference, prefer verifying the user if the authenticator has some form of UV configured.

LDVG replied:
I just realized that this function is used for both credential creation and assertion requests. Note that authenticator behavior is differs for these two operations. You may want to in-line the logic or have two separate functions.

ikerexxe replied:
I think I got it right, but please take a look at it just in case.

> I just realized that this function is used for both credential creation and assertion requests. Note that authenticator behavior differs for these two operations (re: user verification requirements). You may want to in-line the logic or have two separate functions.

No, it's just used for the assertion request.

LDVG replied:
> No, it's just used for the assertion request.

Oh, sorry about that, not sure what I was looking at then.

LDVG commented on src/passkey_child/passkey_child_common.c:
```diff
@@ -319,3 +373,113 @@ register_key(struct passkey_data *data)
+    }
+
+    DEBUG(SSSDBG_TRACE_FUNC, "Resetting assert options.\n");
+    ret = set_assert_options(FIDO_OPT_OMIT, data->user_verification, assert);
```
If the Relying Party has requested user verification (`data->user_verification == FIDO_OPT_TRUE`) and the user has an authenticator with Client PIN configured but without built-in UV support, setting the UV flag to here may cause the authenticator to fail the assertion request. You need to omit the UV flag for such authenticators.

ikerexxe replied:
I added a new case in `get_device_options()` function to handle this.

LDVG commented on src/passkey_child/passkey_child_assert.c:
```diff
@@ -0,0 +1,271 @@
+        goto done;
+    }
+
+    if (_data->type == COSE_ES256) {
```
It looks like the type is expected to be the same over all public keys. What if I have one EDDSA key and one ES256 key?

ikerexxe replied:
I changed what we will print when registering the key and what will be stored in the LDAP server. Nowadays, the public key will be printed and stored in PEM format. This way we can obtain the COSE type from OpenSSL and proceed accordingly.

ikerexxe commented on src/passkey_child/passkey_child_assert.c:
```diff
@@ -0,0 +1,271 @@
+        ret = ERR_INPUT_PARSE;
+        ERROR("PIN required.\n");
+        goto done;
+    }
```
I changed the algorithm to first check for the built-in UV if the key has it. If that fails then it will fallback to PIN authentication.

ikerexxe commented on src/passkey_child/passkey_child_common.c:
```diff
@@ -319,3 +373,113 @@ register_key(struct passkey_data *data)
+        ret = list_devices(dev_list, &dev_list_len);
+        if (ret != EOK) {
+            goto done;
+        }
```
Yes, I moved that out.

ikerexxe commented on src/passkey_child/passkey_child_common.c:
```diff
@@ -319,3 +373,113 @@ register_key(struct passkey_data *data)
+        fido_dev_free(&dev);
+        fido_assert_free(&assert);
+        count++;
+    }
```
I guess you were referring to the operations performed out of the loop and I changed that. If this isn't the case please reopen this point.

ikerexxe commented on src/passkey_child/passkey_child_devices.c:
```diff
@@ -55,32 +55,166 @@ list_devices(fido_dev_info_t *dev_list, size_t *dev_number)
+        if (ret != FIDO_OK) {
+            goto done;
+        }
     }
```
Fixed.

ikerexxe commented:
@LDVG can you take another look at it?

By the way, the CI failures are related to a [libfido2 bug](https://bugzilla.redhat.com/show_bug.cgi?id=2138900). I've already packaged the change and it will be available soon.

LDVG commented on src/passkey_child/passkey_child_devices.c:
```diff
@@ -55,32 +55,190 @@ list_devices(fido_dev_info_t *dev_list, size_t *dev_number)
+              "verification method set.\n");
+        ret = EINVAL;
+        goto done;
+    } else if (_data->user_verification == FIDO_OPT_TRUE && has_uv != true) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "Policy enabled user-verification but the key settings disabled "
+              "it. Thus, omitting it.\n");
+        _data->user_verification = FIDO_OPT_OMIT;
+        ret = EOK;
         goto done;
     }
```
I just had a closer look at the libfido2 behavior: You can safely drop this else-if block if you want to clean up your code slightly. Since you're guarding the built-in UV assertion request with with `fido_dev_has_uv()`, libfido2 will performs this omission for you internally for assertions with a PIN. Sorry for all the back and forth regarding this particularity.

ikerexxe replied:
In that case I remove this part of the code.

ikerexxe commented on src/passkey_child/passkey_child_devices.c:
```diff
@@ -55,32 +55,190 @@ list_devices(fido_dev_info_t *dev_list, size_t *dev_number)
+        _data->user_verification = FIDO_OPT_FALSE;
+        ret = EOK;
         goto done;
     }
```
Once you forced me to reread what I had written it's clear that it doesn't make any sense to have this code so I've removed it.

ikerexxe commented on src/passkey_child/passkey_child_assert.c:
```diff
@@ -0,0 +1,210 @@
+        DEBUG(SSSDBG_OP_FAILURE, "fido_dev_get_assert failed [%d]: %s.\n",
+              ret, fido_strerr(ret));
+        goto done;
+    }
```
>     * The exact same assertion request will be sent to the authenticator twice if `has_uv=true`, `has_pin=false`, `data->user_verification != FIDO_OPT_FALSE`, and the first `fido_dev_get_assert()` fails for some reason. While this is not necessarily a problem, it might be unexpected.

You are right so I added an extra check for it and if there isn't any PIN to fallback then the process fails.

>     * If `has_uv=false`, `has_pin=true`, and `data->user_verification=FIDO_OPT_OMIT`, then the call to `fido_assert_verify()` will not check whether the UV flag was set in the authenticator data. If you want it to be checked, `fido_assert_set_uv(assert, FIDO_OPT_TRUE)` must be called at some point after the second `fido_dev_get_assert()`.
> 
> 
> In general, UV handling might be more clear/concise if it's all done in the same place as the assertion request.

I added a new call to `fido_assert_set_uv()` after calling `fido_dev_get_assert()` to take it into account.

LDVG commented on src/passkey_child/passkey_child_common.c:
```diff
@@ -319,3 +376,181 @@ register_key(struct passkey_data *data)
+    }
+
+    DEBUG(SSSDBG_TRACE_FUNC, "Resetting assert options.\n");
+    ret = set_assert_options(FIDO_OPT_OMIT, data->user_verification, assert);
```
You probably want to set UP to `FIDO_OPT_TRUE` here. Otherwise, `fido_assert_verify()` will not verify that the user presence flag was set in the returned authenticator data.

ikerexxe replied:
Done

ikerexxe commented:
@LDVG do you have any other comments? If not, could you approve the PR?

LDVG commented on src/passkey_child/passkey_child_devices.c:
```diff
@@ -55,32 +55,173 @@ list_devices(fido_dev_info_t *dev_list, size_t *dev_number)
+        DEBUG(SSSDBG_OP_FAILURE, "No device found. Aborting.\n");
+        ret = ENOENT;
         goto done;
+    } else if (dev_list_len == 1) {
```
If there is only one device, no pre-flights are performed. This means that if there's a single authenticator connected to the host which does not contain any known credentials, there will be at least one assertion request fired (two if the device has both built-in UV and PIN configured). While not a blocker (the two assertion requests will both fail with `FIDO_ERR_NO_CREDENTIALS`) it is a distinct difference in behavior from the case where multiple authenticators are plugged in and they are all checked whether they contain any known credential before firing off the actual assertion request.

ikerexxe replied:
As this function is used by the registration and the authentication problem I had to change it a little bit to offer the same behaviour.

LDVG replied:
LGTM! :+1:

LDVG reviewed (approved):
There are outstanding [concerns about PIN handling and authenticator selection per an older thread above](https://github.com/SSSD/sssd/pull/6384#discussion_r1000359007) that I believe to be security critical and detrimental to user experience. I also added a nitpick regarding an edge case on authenticator selection. However, I will not block this PR on either issues if you'd prefer to merge it as-is now and fix these separately.

justin-stephenson commented:
I tried to test the latest code in this PR on F36 using `libfido2-1.10.0-5.fc36.x86_64` but I get an error when running the authenticate command, not sure if it is a problem on my end? This was after resetting the device and adding a PIN.

~~~
[justin@fedora x86_64]$ /usr/libexec/sssd/passkey_child --register --username=justin --domain=test.com                                                                                         
Enter PIN:                                                                                                             
Please touch the device.                                                                                               
passkey:S2G3fgSboyH67n3obhkB0hNnW4+TDHVGEiuW79XXpvOgDtiGh10HhdZTVUiPjr7LrHnuzP7uKslhUGwjFngzbg==,zw8riFHdmli7/JTudrk6fNNPj3qJC7h7VPsDx9iFYhGDrh1ZuRdJdffTbFrIKKtyUvpiQTJOetaAJwiwUXt2Dw==

[justin@fedora x86_64]$ echo 1234 | /usr/libexec/sssd/passkey_child --authenticate --username=justin --domain=test.com --key-handle=S2G3fgSboyH67n3obhkB0hNnW4+TDHVGEiuW79XXpvOgDtiGh10HhdZTVUiPjr7LrHnuzP7uKslhUGwjFngzbg== --public-key=zw8r
iFHdmli7/JTudrk6fNNPj3qJC7h7VPsDx9iFYhGDrh1ZuRdJdffTbFrIKKtyUvpiQTJOetaAJwiwUXt2Dw== --debug-level=9 --logger=stderr
[passkey_child[88126]] [main] (0x0400): passkey_child started.
[passkey_child[88126]] [check_arguments] (0x0400): Argument values after parsing
[passkey_child[88126]] [check_arguments] (0x0400): action: 2
[passkey_child[88126]] [check_arguments] (0x0400): shortname: justin, domain: test.com
[passkey_child[88126]] [check_arguments] (0x0400): Number of keys 1
[passkey_child[88126]] [check_arguments] (0x0400): key 1, public_key: zw8riFHdmli7/JTudrk6fNNPj3qJC7h7VPsDx9iFYhGDrh1ZuRdJdffTbFrIKKtyUvpiQTJOetaAJwiwUXt2Dw==, key_handle: S2G3fgSboyH67n3obhkB0hNnW4+TDHVGEiuW79XXpvOgDtiGh10HhdZTVUiPjr7LrH
nuzP7uKslhUGwjFngzbg==
[passkey_child[88126]] [check_arguments] (0x0400): type: -7 
[passkey_child[88126]] [check_arguments] (0x0400): user_verification: 0
[passkey_child[88126]] [check_arguments] (0x0400): debug_libfido2: 0
[passkey_child[88126]] [authenticate] (0x0400): Checking for devices.
[passkey_child[88126]] [list_devices] (0x0400): Device found.
[passkey_child[88126]] [authenticate] (0x0400): 1 key handles provided.
[passkey_child[88126]] [authenticate] (0x0400): Preparing assert request data with key handle 1.
[passkey_child[88126]] [authenticate] (0x0400): Preparing assert request data.
[passkey_child[88126]] [authenticate] (0x0400): Selecting device.
[passkey_child[88126]] [select_from_multiple_devices] (0x0400): Working with 1 authenticator(s).
[passkey_child[88126]] [select_from_multiple_devices] (0x0200): Assertion found in passkey 0.
[passkey_child[88126]] [authenticate] (0x0400): Comparing the device and policy options.
[passkey_child[88126]] [get_device_options] (0x0100): Policy didn't indicate any preference for user-verification but the key settings are enforcing it. Thus, enforcing the user-verification.
[passkey_child[88126]] [authenticate] (0x0400): Resetting assert options.
[passkey_child[88126]] [authenticate] (0x0400): Resetting assert client data.
[passkey_child[88126]] [authenticate] (0x0400): Decoding public key.
[passkey_child[88126]] [public_key_to_libfido2] (0x0040): d2i_pubkey failed [109052072][error:068000A8:asn1 encoding routines::wrong tag].
Authentication error.
~~~

sumit-bose commented:
> I tried to test the latest code in this PR on F36 using `libfido2-1.10.0-5.fc36.x86_64` but I get an error when running the authenticate command, not sure if it is a problem on my end? This was after resetting the device and adding a PIN.
> 
> ```
> [justin@fedora x86_64]$ /usr/libexec/sssd/passkey_child --register --username=justin --domain=test.com                                                                                         
> Enter PIN:                                                                                                             
> Please touch the device.                                                                                               
> passkey:S2G3fgSboyH67n3obhkB0hNnW4+TDHVGEiuW79XXpvOgDtiGh10HhdZTVUiPjr7LrHnuzP7uKslhUGwjFngzbg==,zw8riFHdmli7/JTudrk6fNNPj3qJC7h7VPsDx9iFYhGDrh1ZuRdJdffTbFrIKKtyUvpiQTJOetaAJwiwUXt2Dw==
> 
> [justin@fedora x86_64]$ echo 1234 | /usr/libexec/sssd/passkey_child --authenticate --username=justin --domain=test.com --key-handle=S2G3fgSboyH67n3obhkB0hNnW4+TDHVGEiuW79XXpvOgDtiGh10HhdZTVUiPjr7LrHnuzP7uKslhUGwjFngzbg== --public-key=zw8r
> iFHdmli7/JTudrk6fNNPj3qJC7h7VPsDx9iFYhGDrh1ZuRdJdffTbFrIKKtyUvpiQTJOetaAJwiwUXt2Dw== --debug-level=9 --logger=stderr
> [passkey_child[88126]] [main] (0x0400): passkey_child started.
> [passkey_child[88126]] [check_arguments] (0x0400): Argument values after parsing
> [passkey_child[88126]] [check_arguments] (0x0400): action: 2
> [passkey_child[88126]] [check_arguments] (0x0400): shortname: justin, domain: test.com
> [passkey_child[88126]] [check_arguments] (0x0400): Number of keys 1
> [passkey_child[88126]] [check_arguments] (0x0400): key 1, public_key: zw8riFHdmli7/JTudrk6fNNPj3qJC7h7VPsDx9iFYhGDrh1ZuRdJdffTbFrIKKtyUvpiQTJOetaAJwiwUXt2Dw==, key_handle: S2G3fgSboyH67n3obhkB0hNnW4+TDHVGEiuW79XXpvOgDtiGh10HhdZTVUiPjr7LrH
> nuzP7uKslhUGwjFngzbg==
> [passkey_child[88126]] [check_arguments] (0x0400): type: -7 
> [passkey_child[88126]] [check_arguments] (0x0400): user_verification: 0
> [passkey_child[88126]] [check_arguments] (0x0400): debug_libfido2: 0
> [passkey_child[88126]] [authenticate] (0x0400): Checking for devices.
> [passkey_child[88126]] [list_devices] (0x0400): Device found.
> [passkey_child[88126]] [authenticate] (0x0400): 1 key handles provided.
> [passkey_child[88126]] [authenticate] (0x0400): Preparing assert request data with key handle 1.
> [passkey_child[88126]] [authenticate] (0x0400): Preparing assert request data.
> [passkey_child[88126]] [authenticate] (0x0400): Selecting device.
> [passkey_child[88126]] [select_from_multiple_devices] (0x0400): Working with 1 authenticator(s).
> [passkey_child[88126]] [select_from_multiple_devices] (0x0200): Assertion found in passkey 0.
> [passkey_child[88126]] [authenticate] (0x0400): Comparing the device and policy options.
> [passkey_child[88126]] [get_device_options] (0x0100): Policy didn't indicate any preference for user-verification but the key settings are enforcing it. Thus, enforcing the user-verification.
> [passkey_child[88126]] [authenticate] (0x0400): Resetting assert options.
> [passkey_child[88126]] [authenticate] (0x0400): Resetting assert client data.
> [passkey_child[88126]] [authenticate] (0x0400): Decoding public key.
> [passkey_child[88126]] [public_key_to_libfido2] (0x0040): d2i_pubkey failed [109052072][error:068000A8:asn1 encoding routines::wrong tag].
> Authentication error.
> ```

Hi,

I think this is because of the different key types. The latest code expects the base64 PEM string for verification, while --register in this PR still outputs the plain key format. @ikerexxe, can you rebase the PR on top of current master?

bye,
Sumit

ikerexxe commented:
> I think this is because of the different key types. The latest code expects the base64 PEM string for verification, while --register in this PR still outputs the plain key format. @ikerexxe, can you rebase the PR on top of current master?

Sumit is right. Now that the code is rebased you should be able to generate a key and verify it.

sumit-bose commented:
Hi,

I played with the options a bit. If a wrong domain (relying party id) is given with `--domain` authentication fails. But it looks like `--username` is completely unused. The name is stored during the registration (at least when using discoverable credentials) because it is shown e.g. with `ykman fido credentials list`. But it is not used during verification and it looks that typically it will not even be returned with the assertion as documented in https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-20210615.html#authenticatorgetassertion-response-structure. And if you follow the definition of `PublicKeyCredentialUserEntity` it looks like only the display name is returned together with the `id`.

On the other hand there is the `fido_assert_user_name()` which implies that there might be cases where the user name is present in the assertion.

Additionally, according to the spec even the `id` is only returned for discoverable credentials and if you follow the link for `server side credentials`  aka `non-resident credentials` it says that those are deprecated.

With this I have a couple of questions, maybe @LDVG has some answers:
 - would it make sense to default to discoverable credentials? Currently we use the default of the device which looks is typically the deprecated type.
 - Is there a way to make sure the assertion data contains the user name?
 - If the user name cannot be with the assertion data, would it make sense to use it in the selection of the key together with the key handle?
 -  If the user name cannot be with the assertion data, would it make sense to store the random `id` on the server so that if can be used to verify the assertion?

bye,
Sumit

LDVG commented:
> Additionally, according to the spec even the `id` is only returned for discoverable credentials and if you follow the link for `server side credentials` aka `non-resident credentials` it says that those are deprecated.

Server-side/non-discoverable credentials are not deprecated. I think you are referring to the deprecated _name_ "non-resident credential"? 

> * would it make sense to default to discoverable credentials? Currently we use the default of the device which looks is typically the deprecated type.

This depends mostly on what your implementation goals are. Authenticators can generate assertions for a discoverable credential given only the RP ID. They then additionally return the user ID (aka. user handle) which allows the RP to look up the associated user. This means that the user does not need to enter a username to login, enabling "usernameless" authentication ceremonies. Otherwise, if you are content with identifying the user in some other way, server-side credentials should be sufficient.

> * Is there a way to make sure the assertion data contains the user name?

Per the specification, only discoverable credentials return user account information in the assertion response and it is only guaranteed to contain the user ID.

> * If the user name cannot be with the assertion data, would it make sense to use it in the selection of the key together with the key handle?

I'm not sure I follow. Which selection of what key are we referring to here?

> * If the user name cannot be with the assertion data, would it make sense to store the random `id` on the server so that if can be used to verify the assertion?

You might want to do this, yes. See for example WebAuthn [§ 7.2. Verifying an Authentication Assertion, step 6](https://www.w3.org/TR/webauthn-2/#sctn-verifying-assertion).

sumit-bose commented:
> > Additionally, according to the spec even the `id` is only returned for discoverable credentials and if you follow the link for `server side credentials` aka `non-resident credentials` it says that those are deprecated.
> 
> Server-side/non-discoverable credentials are not deprecated. I think you are referring to the deprecated _name_ "non-resident credential"?

ah, thanks, I misinterpreted the 'deprecated' comment.

> 
> > * would it make sense to default to discoverable credentials? Currently we use the default of the device which looks is typically the deprecated type.
> 
> This depends mostly on what your implementation goals are. Authenticators can generate assertions for a discoverable credential given only the RP ID. They then additionally return the user ID (aka. user handle) which allows the RP to look up the associated user. This means that the user does not need to enter a username to login, enabling "usernameless" authentication ceremonies. Otherwise, if you are content with identifying the user in some other way, server-side credentials should be sufficient.

There is similar "usernameless" authentication feature for Smartcard authentication but it only works when using GDM. In general I think such feature is not needed here. 

> 
> > * Is there a way to make sure the assertion data contains the user name?
> 
> Per the specification, only discoverable credentials return user account information in the assertion and it is only guaranteed to contain the user ID.
> 
> > * If the user name cannot be with the assertion data, would it make sense to use it in the selection of the key together with the key handle?
> 
> I'm not sure I follow. Which selection of what key are we referring to here?
>

I meant the key on the FIDO2 device and iterate oven the keys like e.g. `ykman fido credentials list` does and check if there is a key where key handle and user name matches and only if there is a match continue and call `fido_dev_get_assert()`. But I guess I can answer the question myself and this would not give an additional protection.
 
> > * If the user name cannot be with the assertion data, would it make sense to store the random `id` on the server so that if can be used to verify the assertion?
> 
> You might want to do this, yes. See for example WebAuthn [§ 7.2. Verifying an Authentication Assertion, step 6](https://www.w3.org/TR/webauthn-2/#sctn-verifying-assertion).

Thanks, but then we would need discoverable credentials.

justin-stephenson commented on src/passkey_child/passkey_child_credentials.c:
```diff
@@ -137,6 +139,49 @@ prepare_credentials(struct passkey_data *data, fido_dev_t *dev,
+    }
+
+    /* Remove the end of line '\n' character */
+    len--;
```
Is this needed?  When calling `passkey_child --authenticate...` from SSSD I had to remove this line, or the trailing digit of the PIN was chopped off. I used `echo -n $PIN` when testing on the command line.

ikerexxe replied:
Removed. Thanks for reporting the issue.

sumit-bose commented:
Hi,

given the previous discussion about user name and the assertion data I'd like to make the following suggestions:

 - the `--username` option should not be required for `--authenticate` since it is neither used nor needed
 - there should be a new option for `--register` to choose `server side` (the ones we use currently) or `discoverable` credentials; in the latter case the register output should contain a third field with the random user id; this can be handled in a new PR

If and how the user id is used on the KDC side during the Kerberos authentication can be decided when this is implemented.

bye,
Sumit

ikerexxe commented:
>     * the `--username` option should not be required for `--authenticate` since it is neither used nor needed

Done.

>     * there should be a new option for `--register` to choose `server side` (the ones we use currently) or `discoverable` credentials; in the latter case the register output should contain a third field with the random user id; this can be handled in a new PR

I'll work on this and open a PR next week.

Please approve if you agree with the current PR state.

justin-stephenson reviewed (approved):
Testing authentication from SSSD is working as expected, aside from some passkey_child SELinux denials. I will investigate these further but Ack from me.

sumit-bose reviewed (approved):
Hi,

thanks, ACK.

bye,
Sumit

pbrezina commented:
Pushed PR: https://github.com/SSSD/sssd/pull/6384

* `master`
    * 6b0d175f8fad63c9c41e0b990c4a5632e96cc3f7 - passkey: input PIN via stdin
    * 2b0a8f2757c013ecdaad2047f6254f2ef3794cb4 - passkey: public key in PEM format
    * ab6910ae0e4e32692cbab357129a4bde8d2217bf - passkey: verify assertion
